app/Models/Device.php

```diff
@@ -324,15 +324,33 @@ public function getAttribs()
         return $this->attribs->pluck('attrib_value', 'attrib_type')->toArray();
     }
 
-    public function setLocation($location_text)
+    /**
+     * Update the location to the correct location and update GPS if needed
+     *
+     * @param  \App\Models\Location  $location location data
+     */
+    public function setLocation(Location $location)
     {
-        $location_text = $location_text ? Rewrite::location($location_text) : null;
+        $location->location = $location->location ? Rewrite::location($location->location) : null;
 
-        $this->location_id = null;
-        if ($location_text) {
-            $location = Location::firstOrCreate(['location' => $location_text]);
+        if (! $location->location) {
+            $this->location_id = null;
+
+            return;
+        }
+
+        $coord = array_filter($location->only(['lat', 'lng']));
+        if (! $this->relationLoaded('location') || optional($this->location)->location !== $location->location) {
```

Comment:
I must be missing something here,
in what cases would the location relation be loaded already? and why would it matter for this?

Reply:
If custom os code loads the relation, we want to use that instead of fetching it from the DB.
Or if the new location is already saved to the DB, we can just use that.

Reply:
I tend to think of code that normal contributors will write as a barrier.  Anything can happen outside that barrier :)